pkg/kv/kvserver/replica_proposal_buf.go

```diff
@@ -456,61 +461,170 @@ func (b *propBuf) FlushLockedWithRaftGroup(
+		// conditional should be encapsulated in methods in the `logstore` package.
+		if entryEncoding == raftlog.EntryEncodingStandard || entryEncoding == raftlog.EntryEncodingSideloaded {
+			// Allocate the data slice with just the right capacity. This is a hot
+			// path so we are careful to minimize allocations.
+			if ct.IsSet() {
+				// See assignment to `ct` above to understand ctPool. It's important that we
+				// do this before calling p.command.Size() below or that will be off.
+				ctp := ctPool.Get().(*hlc.Timestamp)
+				*ctp = ct
+				p.command.ClosedTimestamp = ctp
+			}
+			data := make([]byte, raftlog.RaftCommandPrefixLen+p.command.Size())
+			// Encode prefix with command ID.
+			prefixByte := raftlog.EntryEncodingStandardPrefixByte
+			if entryEncoding == raftlog.EntryEncodingSideloaded {
+				prefixByte = raftlog.EntryEncodingSideloadedPrefixByte
+			}
+			raftlog.EncodeRaftCommandPrefix(data[:raftlog.RaftCommandPrefixLen], prefixByte, p.idKey)
+			// Encode body of command into the remaining space, which it will fill precisely.
+
+			if _, err := protoutil.MarshalTo(
+				p.command, data[raftlog.RaftCommandPrefixLen:],
+			); err != nil {
 				firstErr = err
 				continue
 			}
-		}
+			if ctp := p.command.ClosedTimestamp; ctp != nil {
+				p.command.ClosedTimestamp = nil
+				ctPool.Put(ctp)
+			}
 
-		// Coordinate proposing the command to etcd/raft.
-		if crt := p.command.ReplicatedEvalResult.ChangeReplicas; crt != nil {
-			// Flush any previously batched (non-conf change) proposals to
-			// preserve the correct ordering or proposals. Later proposals
-			// will start a new batch.
-			if err := proposeBatch(raftGroup, b.p.getReplicaID(), ents); err != nil {
-				firstErr = err
+			if drop, err := testingSubmitProposalFilter(p); drop || err != nil {
+				firstErr = errors.CombineErrors(firstErr, err)
 				continue
 			}
-			ents = ents[len(ents):]
 
+			// Add to the batch of entries that will soon be proposed. It is
+			// possible that this batching can cause the batched MsgProp to grow
+			// past the size limit where etcd/raft will drop the entire thing
+			// (see raft.Config.MaxUncommittedEntriesSize), but that's not a
+			// concern. This setting is configured to twice the maximum quota in
+			// the proposal quota pool, so for batching to cause a message to be
+			// dropped the uncommitted portion of the Raft log would already
+			// need to be at least as large as the proposal quota size, assuming
+			// that all in-flight proposals are reproposed in a single batch.
+
+			// Log an event if this is a large proposal. These are more likely to cause
+			// blips or worse, and it's good to be able to pick them from traces.
+			const largeProposalEventThresholdBytes = 2 << 19 // 512kb
+			if n := len(data); n > largeProposalEventThresholdBytes {
+				log.Eventf(p.ctx, "proposal is large: %s", humanizeutil.IBytes(int64(n)))
+			}
+
+			ents = append(ents, raftpb.Entry{
+				Data: data,
+			})
+		} else {
```

Comment:
With the `if-else` construction it is assumed that if the encoding isn't one of the 2 in `if` branch, then it's a config change. This is error-prone if any new encoding is introduced. How about spelling out the encoding explicitly? Maybe use `switch-case` instead. This would also make the first line of the comment below unnecessary.